src/mlpack/methods/kde/kde_rules_impl.hpp

```diff
@@ -0,0 +1,194 @@
+/**
+ * @file kde_rules_impl.hpp
+ * @author Roberto Hueso (robertohueso96@gmail.com)
+ *
+ * Implementation of rules for Kernel Density Estimation with generic trees.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+
+#ifndef MLPACK_METHODS_KDE_RULES_IMPL_HPP
+#define MLPACK_METHODS_KDE_RULES_IMPL_HPP
+
+// In case it hasn't been included yet.
+#include "kde_rules.hpp"
+
+namespace mlpack {
+namespace kde {
+
+template<typename MetricType, typename KernelType, typename TreeType>
+KDERules<MetricType, KernelType, TreeType>::KDERules(
+    const arma::mat& referenceSet,
+    const arma::mat& querySet,
+    arma::vec& densities,
+    const double relError,
+    const double absError,
+    const std::vector<size_t>& oldFromNewQueries,
+    MetricType& metric,
+    KernelType& kernel) :
+    referenceSet(referenceSet),
+    querySet(querySet),
+    densities(densities),
+    absError(absError),
+    relError(relError),
+    oldFromNewQueries(oldFromNewQueries),
+    metric(metric),
+    kernel(kernel),
+    lastQueryIndex(querySet.n_cols),
+    lastReferenceIndex(referenceSet.n_cols),
+    baseCases(0),
+    scores(0)
+{
+  // Nothing to do.
+}
+
+//! The base case.
+template<typename MetricType, typename KernelType, typename TreeType>
+inline force_inline
+double KDERules<MetricType, KernelType, TreeType>::BaseCase(
+    const size_t queryIndex,
+    const size_t referenceIndex)
+{
+  double distance = metric.Evaluate(querySet.col(queryIndex),
+                                    referenceSet.col(referenceIndex));
+  if (tree::TreeTraits<TreeType>::RearrangesDataset)
+    densities(oldFromNewQueries.at(queryIndex)) += kernel.Evaluate(distance);
+  else
+    densities(queryIndex) += kernel.Evaluate(distance);
+  ++baseCases;
+  lastQueryIndex = queryIndex;
+  lastReferenceIndex = referenceIndex;
+  return distance;
+}
+
+//! Single-tree scoring function.
+template<typename MetricType, typename KernelType, typename TreeType>
+double KDERules<MetricType, KernelType, TreeType>::
+Score(const size_t /* queryIndex */, TreeType& /* referenceNode */)
+{
+  ++scores;
+  traversalInfo.LastScore() = 0.0;
+  return 0.0;
```

Comment:
Ah, I think maybe this could cause a problem.  The `BinarySpaceTree` dual-tree traverser (which is used for `KDTree` and `BallTree`) uses a call to the single-tree `Score()` after calling the dual-tree `Score()`, once it hits a query leaf node.  But I can see that the tests are passing, so I am a little confused on why the code works despite this.  I'll have to look into it, unless you already know the reason.

Reply:
Fixed in d023b8266be0d9e43d64dfd1f8d0e2b00ab0db4b
It worked before because all leaf nodes were visited.